src/implicit.rs

```diff
@@ -193,20 +199,21 @@ pub fn resolve_neutral(sequence: &IsolatingRunSequence, levels: &[u8], classes:
 /// Returns the maximum embedding level in the paragraph.
 ///
 /// http://www.unicode.org/reports/tr9/#Resolving_Implicit_Levels
-pub fn resolve_levels(classes: &[BidiClass], levels: &mut [u8]) -> u8 {
-    let mut max_level = 0;
+pub fn resolve_levels(classes: &[BidiClass], levels: &mut [Level]) -> Level {
+    let mut max_level = Level::ltr();
 
     assert!(classes.len() == levels.len());
     for i in 0..levels.len() {
-        match (is_rtl(levels[i]), classes[i]) {
+        match (levels[i].is_rtl(), classes[i]) {
             // http://www.unicode.org/reports/tr9/#I1
-            (false, R) => levels[i] += 1,
-            (false, AN) | (false, EN) => levels[i] += 2,
+            (false, R) => levels[i].raise(1).expect("Level number error"),
+            (false, AN) | (false, EN) => levels[i].raise(2).expect("Level number error"),
             // http://www.unicode.org/reports/tr9/#I2
-            (true, L) | (true, EN) | (true, AN) => levels[i] += 1,
+            (true, L) | (true, EN) | (true, AN) => levels[i].raise(1).expect("Level number error"),
```

Comment:
According to the [spec], the level may reach *max_depth* + 1 at this step, so we need different validity checking here than in earlier stages.

[spec]: http://www.unicode.org/reports/tr9/#Resolving_Implicit_Levels